Add script to build ocamlformat with mingw64

nojb opened the pull request:
I noticed that this project has a need for easier testing and building on Windows (see #49 and #1700 and the issues linked therein). 

To try to improve the situation I wrote a quick-and-dirty shell script which will build a native Windows binary (using `mingw64` compiler) when run inside a standard Cygwin installation.

Since the script takes care of downloading and building `ocaml`, `flexlink`, `dune`, and all its dependencies, only the base `cygwin` system is required (with the `mingw64` compiler, of course!).

~~The exact version of each package is hard-coded in the script so it may need tweaking from time to time, but it sure beats the current situation (no easy way to build on Windows at all)!~~ (**UPDATE: we now use OPAM instead of manually specifying the dependencies.**)

Cheers!

nojb commented:
The second commit adds a GH action that can be triggered on-demand to build the binary using the script. The resulting artifact is uploaded back into GitHub. It takes ~20 minutes to complete. The action can be used to build an arbitrary commit, and it will use the version of the script present at that commit.

You can see a full log of the action here: https://github.com/nojb/ocamlformat/runs/2981199315?check_suite_focus=true

Triggering the workflow:

![image](https://user-images.githubusercontent.com/113560/124369109-9f2a6f80-dc68-11eb-9215-b4c5179a6b29.png)

Uploading the artifact:

![image](https://user-images.githubusercontent.com/113560/124370182-b1120f80-dc74-11eb-93af-f8e7bed2aa41.png)

Downloading the artifact:

![image](https://user-images.githubusercontent.com/113560/124370190-c8e99380-dc74-11eb-80ae-d4823ee26339.png)

emillon commented:
FYI, the current approach I use to create windows binaries (using opam) is documented in `HACKING.org`. I think it is very nice to have an automated way to do this (though the endgame is for ocaml-ci to do that).

emillon commented on tools/build-mingw64.sh:
```diff
@@ -0,0 +1,104 @@
+
+packages=(
+    https://github.com/janestreet/sexplib0                v0.14.0
+    https://github.com/ocaml-dune/csexp                   master
```
we're not terribly reproducible because of transitive deps, but I think we should avoid refering to `master` if possible

emillon commented on tools/build-mingw64.sh:
```diff
@@ -0,0 +1,104 @@
+
+cd ..
+
+dune build ocamlformat.install
```
I think this creates a build using the dev profile. In `HACKING.org` there are some instructions to create a release binary using `dune-release`, maybe it can be used here?  (not sure with the intermediate clone step)

nojb commented:
> FYI, the current approach I use to create windows binaries (using opam) is documented in `HACKING.org`. I think it is very nice to have an automated way to do this (though the endgame is for ocaml-ci to do that).

Thanks for the review. I took a quick look at `HACKING.org` and didn't see any documentation on how to setup OPAM on native Windows. Is this documented anywhere?

emillon commented:
That part is not in the document, indeed. The steps are:
- [Install OCaml for windows](https://fdopen.github.io/opam-repository-mingw/installation/) - not sure how easy it is to do headless
- `git clone` ocamlformat and cd into it
- `opam switch create ./ 4.12.0+mingw64c` (will download a binary version of the compiler without compiling it)
- then you can install and run `dune-release distrib`

nojb commented:
> That part is not in the document, indeed. The steps are:

Thanks for the hint.

I updated the PR to use `opam` which is much better. I could not use `dune-release` because it seems to have trouble quoting things for a Cygwin-compiled `git`, but this is only for the last step so I just replaced that by a direct call to `dune`.

nojb commented:
Friendly ping.

gpetiot commented:
> I updated the PR to use `opam` which is much better. I could not use `dune-release` because it seems to have trouble quoting things for a Cygwin-compiled `git`, but this is only for the last step so I just replaced that by a direct call to `dune`.

About that please don't hesitate to open an issue on `dune-release`

gpetiot commented:
I don't have the cygwin setup to test this branch, @emillon is everything working for you?

nojb commented:
> I don't have the cygwin setup to test this branch, @emillon is everything working for you?

Note that this is (was) tested via the Github action https://github.com/nojb/ocamlformat/actions/runs/1006853536

gpetiot reviewed (approved):
Looks good to me, thanks!

nojb commented:
Thanks!

Psoffie009 commented:
> I noticed that this project has a need for easier testing and building on Windows (see #49 and #1700 and the issues linked therein).
> 
> To try to improve the situation I wrote a quick-and-dirty shell script which will build a native Windows binary (using `mingw64` compiler) when run inside a standard Cygwin installation.
> 
> Since the script takes care of downloading and building `ocaml`, `flexlink`, `dune`, and all its dependencies, only the base `cygwin` system is required (with the `mingw64` compiler, of course!).
> 
> ~The exact version of each package is hard-coded in the script so it may need tweaking from time to time, but it sure beats the current situation (no easy way to build on Windows at all)!~ (**UPDATE: we now use OPAM instead of manually specifying the dependencies.**)
> 
> Cheers!

Psoffie009 commented:
> Hi @nojb!
> 
> Thank you for your pull request and welcome to our community.
> 
> # Action Required
> In order to merge **any pull request** (code, docs, etc.), we **require** contributors to sign our **Contributor License Agreement**, and we don't seem to have one on file for you.
> 
> # Process
> In order for us to review and merge your suggested changes, please sign at https://code.facebook.com/cla. **If you are contributing on behalf of someone else (eg your employer)**, the individual CLA may not be sufficient and your employer may need to sign the corporate CLA.
> 
> Once the CLA is signed, our tooling will perform checks and validations. Afterwards, the **pull request will be tagged** with `CLA signed`. The tagging process may take up to 1 hour after signing. Please give it that time before contacting us about it.
> 
> If you have received this in error or have any questions, please contact us at [cla@fb.com](mailto:cla@fb.com?subject=CLA%20for%20ocaml-ppx%2Focamlformat%20%231702). Thanks!

Psoffie009 commented:
> The second commit adds a GH action that can be triggered on-demand to build the binary using the script. The resulting artifact is uploaded back into GitHub. It takes ~15 minutes to complete. The action can be used to build an arbitrary commit, and it will use the version of the script present at that commit.
> 
> You can see a full log of the action here: https://github.com/nojb/ocamlformat/runs/2981199315?check_suite_focus=true
> 
> ![image](https://user-images.githubusercontent.com/113560/124369109-9f2a6f80-dc68-11eb-9215-b4c5179a6b29.png)
> 
> ![image](https://user-images.githubusercontent.com/113560/124391770-6fc04500-dcf2-11eb-989d-bdbb5d5170ec.png)
> 
> ![image](https://user-images.githubusercontent.com/113560/124391786-7ea6f780-dcf2-11eb-8feb-b7f815a70cde.png)
